### PR TITLE
fix: hashmap.huff hashing pattern changed to solidity pattern

### PR DESCRIPTION
**Pull request by AmadiMichael**
This changes the pattern of hashing for slots in `Hashmap.huff` to match that of how solidity lang handles mappings.

`GET_SLOT_FROM_KEYS()` function uses `sha3(slot . key)` against `sha3(key . slot)` used in solidity,
`GET_SLOT_FROM_KEYS_2D()` function uses `sha3(sha3(slot . key1) . key2)` against `sha3(key2 . sha3(key1 . slot)` used in solidity and 
`GET_SLOT_FROM_KEYS_3D()` function uses `sha3(key3 . sha3(key2 . sha3(slot . key1)))` against `sha3(key3 . sha3(key2 . sha3(key1 . slot)))` used in solidity.

### CODE REVIEW

**Comment by devtooligan** [comment]
lgtm. would be nice to get a second set of eyes on this one since i was involved in the creation of this fix @cheethas @clabby @refcell @MathisGD

File: src/data-structures/Hashmap.huff [hunk 1]
```diff
@@ -17,14 +17,13 @@
...
+    // Input stack: [slot, key2]
     // Load the data into memory.
-    <mem_ptr>           // [<mem_ptr>, key1, key2]
+    <mem_ptr> 0x20 add  // [<mem_ptr> + 32, slot, key2]
     mstore              // [key2]
-    <mem_ptr> 0x20 add  // [<mem_ptr> + 32, key2]
+    <mem_ptr>           // [<mem_ptr>, key2]
     mstore              // []
 
     // Hash the data, generating a key.
```
Comment (MathisGD) [inline src/data-structures/Hashmap.huff:29]:
```suggestion
    // Hash the data, generating a slot.
```
same issue line 32

**Review by MathisGD — APPROVED** [review]
Minor issues with comments, otherwise LGTM

**Comment by AmadiMichael** [comment]
> Minor issues with comments, otherwise LGTM

Ah, missed them, thanks.
Committed.

File: src/data-structures/Hashmap.huff [hunk 2]
```diff
@@ -17,17 +17,16 @@
...
     mstore              // [key2]
-    <mem_ptr> 0x20 add  // [<mem_ptr> + 32, key2]
+    <mem_ptr>           // [<mem_ptr>, key2]
     mstore              // []
 
-    // Hash the data, generating a key.
+    // Hash the data, generating a slot.
     0x40        // [64]
     <mem_ptr>   // [<mem_ptr>, 64]
     sha3        // [key]
```
Comment (MathisGD) [inline src/data-structures/Hashmap.huff:32]:
```suggestion
    sha3        // [slot]
```

File: src/data-structures/Hashmap.huff [hunk 3]
```diff
@@ -17,17 +17,16 @@
...
+/// @notice Given two keys (ie a slot and a key), hash them together, generating a slot for a secondary hashmap.
 #define macro GET_SLOT_FROM_KEYS(mem_ptr) = takes(2) returns (1) {
-    // Input stack: [key1, key2]
+    // Input stack: [slot, key]
     // Load the data into memory.
-    <mem_ptr>           // [<mem_ptr>, key1, key2]
+    <mem_ptr> 0x20 add  // [<mem_ptr> + 32, slot, key2]
     mstore              // [key2]
-    <mem_ptr> 0x20 add  // [<mem_ptr> + 32, key2]
+    <mem_ptr>           // [<mem_ptr>, key2]
```
Comment (MathisGD) [inline src/data-structures/Hashmap.huff:26]:
```suggestion
    <mem_ptr>           // [<mem_ptr>, key]
```

File: src/data-structures/Hashmap.huff [hunk 4]
```diff
@@ -17,17 +17,16 @@
...
 
-/// @notice Given two keys (ie an address and a number), hash them together, generating a slot for a secondary hashmap.
-///         This should only be used if you have multiple maps in your contract.
+/// @notice Given two keys (ie a slot and a key), hash them together, generating a slot for a secondary hashmap.
 #define macro GET_SLOT_FROM_KEYS(mem_ptr) = takes(2) returns (1) {
-    // Input stack: [key1, key2]
+    // Input stack: [slot, key]
     // Load the data into memory.
-    <mem_ptr>           // [<mem_ptr>, key1, key2]
+    <mem_ptr> 0x20 add  // [<mem_ptr> + 32, slot, key2]
```
Comment (MathisGD) [inline src/data-structures/Hashmap.huff:24]:
```suggestion
    <mem_ptr> 0x20 add  // [<mem_ptr> + 32, slot, key]
```

File: src/data-structures/Hashmap.huff [hunk 5]
```diff
@@ -17,17 +17,16 @@
...
     mstore              // [key2]
```
Comment (MathisGD) [inline src/data-structures/Hashmap.huff:25]:
```suggestion
    mstore              // [key]
```

**Review by MathisGD — COMMENTED** [review]
Last comments. Note that Github allows you to batch the suggestions into one commit.

**Review by MathisGD — APPROVED** [review]

**Review by devtooligan — APPROVED** [review]
🚀

**Review by refcell — APPROVED** [review]

**Pull request merged**